feat(config): add and apply cluster change operation for delete exporter

giampa91 opened the pull request:
## Description

Implements:

- #35322: Add PartitionDeleteExporterOperation to represent a request to delete an exporter from a partition
- #35323: Add ClusterChangeOperationApplier to apply delete-exporter operations to partition configurations

Details:

- Introduce a new `PartitionDeleteExporterOperation` in the cluster configuration model
- Implement `ExporterDeleteRequestTransformer` to generate delete operations for all relevant partitions
- Update `ClusterChangePlanExecutor` and `PartitionChangeExecutor` to support the new operation type
- Apply config mutation in `PartitionDeleteExporterOperationApplier`
- Add tests to validate transformer behavior and applier logic

## Checklist

## Related issues

closes #35322 #35323

deepthidevaki commented on zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/changes/PartitionDeleteExporterApplier.java:
```diff
@@ -0,0 +1,109 @@
+import io.camunda.zeebe.util.Either;
+import java.util.function.UnaryOperator;
+
+final class PartitionDeleteExporterApplier implements MemberOperationApplier {
```
:x: We need a test for this class that verifies following behaviors:

* Init fails when
    - If exporter state is not `CONFIG_NOT_FOUND`
    - no member exists
    - no partition exists
    - no exporter exists
* Apply returns failed future when `partitionChangeExecutor` fails
* Success scenario: The exporter does not exist in config, after the operation is applied

deepthidevaki commented on zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/changes/PartitionDeleteExporterApplier.java:
```diff
@@ -0,0 +1,109 @@
+
+  @Override
+  public ActorFuture<UnaryOperator<MemberState>> applyOperation() {
+    final var result = new CompletableActorFuture<UnaryOperator<MemberState>>();
+
+    partitionChangeExecutor
+        .deleteExporter(partitionId, exporterId)
+        .onComplete(
+            (nothing, error) -> {
+              if (error == null) {
+                result.complete(
+                    memberState ->
+                        memberState.updatePartition(
+                            partitionId,
+                            partition ->
+                                partition.updateConfig(
+                                    config -> deleteExporter(config, exporterId))));
+              } else {
+                result.completeExceptionally(error);
+              }
+            });
+
+    return result;
```
:wrench: You can simplify is using `ActorFuture#thenApply`
```suggestion
 return partitionChangeExecutor
        .deleteExporter(partitionId, exporterId)
        .thenApply(
            (nothing) -> {
              return memberState ->
                  memberState.updatePartition(
                      partitionId,
                      partition ->
                          partition.updateConfig(config -> deleteExporter(config, exporterId)));
            });
```

deepthidevaki reviewed (changes requested):
Thanks a lot for your contribution. Overall it looks good. I do have some comments. Once you apply these changes, I will enable CI workflow to see if everything passes.

:x:  There is one main thing missing from the PR related to the [second issue](https://github.com/camunda/camunda/issues/35323). We have to implement the actual logic of deleting the exporter from the `ExporterDirector` runtime. I would expect this in a different PR. Please don't add it to this PR. Instead, please add a placeholder implementation in `PartitionManagerImpl` which for now just throws an Unsupported operation exception`, so that the project compiles.

:wrench: In general, it would be good to split such PRs into multiple commits for easier review. For example, here at the least two commits referring to the two issues would have been nice. :slightly_smiling_face:

giampa91 committed:
feat: add cluster configuration management API and request handlers

- Implement core API classes for cluster configuration management
- Add request sender, handler, and server components
- Add exporter delete request transformer

giampa91 committed:
feat: implement configuration change appliers and partition change executors

- Add implementations for configuration change appliers
- Add partition change executors including PartitionDeleteExporterApplier

giampa91 committed:
feat: add serializers for cluster configuration requests

- Implement protobuf-based serializers
- Add cluster configuration requests serializer

giampa91 committed:
feat: add cluster configuration change operations and exporters config

- Define change operation models
- Implement exporters config state objects

giampa91 committed:
feat: add protobuf definitions for cluster configuration requests and topology

giampa91 committed:
test: add unit tests for cluster configuration API, transformers, and serializers, applier

giampa91 committed:
fix: apply consistent code formatting and cleanup

giampa91 committed:
fix: insert PartitionDeleteExporterOperation in ClusterApiUtilsTest

giampa91 commented:
I have updated the pull request with update with rebase @deepthidevaki

deepthidevaki reviewed (approved):
I have opened an issue for the CI failures. https://github.com/camunda/camunda/issues/35922 It is definitely not related to your changes. 

Merging will be blocked until we fix it or find a way around. Sorry. :disappointed:

deepthidevaki commented:
@giampa91 The CI fix is merged. Could you please rebase on main?

deepthidevaki commented:
Looks like the issue fixed by this PR https://github.com/camunda/camunda/pull/36074. Let's wait until this is merged.
